be/src/runtime/sender_queue.cpp

```diff
@@ -0,0 +1,899 @@
+        return;
+    }
+    _sender_eos_set.insert(be_number);
+    DCHECK_GT(_num_remaining_senders, 0);
+    _num_remaining_senders--;
+    VLOG_FILE << "decremented senders: fragment_instance_id=" << print_id(_recvr->fragment_instance_id())
+              << " node_id=" << _recvr->dest_node_id() << " #senders=" << _num_remaining_senders
+              << " be_number=" << be_number;
+    if (_num_remaining_senders == 0) {
+        _data_arrival_cv.notify_one();
+    }
+}
+
+void DataStreamRecvr::NonPipelineSenderQueue::cancel() {
+    {
+        std::lock_guard<Mutex> l(_lock);
+        if (_is_cancelled) {
+            return;
+        }
+        _is_cancelled = true;
+        VLOG_QUERY << "cancelled stream: _fragment_instance_id=" << _recvr->fragment_instance_id()
+                   << " node_id=" << _recvr->dest_node_id();
+    }
+    // Wake up all threads waiting to produce/consume batches.  They will all
+    // notice that the stream is cancelled and handle it.
+    _data_arrival_cv.notify_all();
+
+    {
+        std::lock_guard<Mutex> l(_lock);
+        clean_buffer_queues();
+    }
+}
+
+void DataStreamRecvr::NonPipelineSenderQueue::close() {
+    {
+        // If _is_cancelled is not set to true, there may be concurrent send
+        // which add batch to _batch_queue. The batch added after _batch_queue
+        // is clear will be memory leak
+        std::lock_guard<Mutex> l(_lock);
+        _is_cancelled = true;
+
+        clean_buffer_queues();
+    }
+}
+
+void DataStreamRecvr::NonPipelineSenderQueue::clean_buffer_queues() {
+    for (auto& item : _chunk_queue) {
+        if (item.closure != nullptr) {
+            item.closure->Run();
+        }
+    }
+    _chunk_queue.clear();
+    for (auto& [_, chunk_queues] : _buffered_chunk_queues) {
+        for (auto& [_, chunk_queue] : chunk_queues) {
+            for (auto& item : chunk_queue) {
+                if (item.closure != nullptr) {
+                    item.closure->Run();
+                }
+            }
+        }
+    }
+    _buffered_chunk_queues.clear();
+}
+
+DataStreamRecvr::PipelineSenderQueue::PipelineSenderQueue(DataStreamRecvr* parent_recvr, int32_t num_senders,
+                                                          int32_t degree_of_parallism)
+        : SenderQueue(parent_recvr), _num_remaining_senders(num_senders) {
+    for (int i = 0; i < degree_of_parallism; i++) {
+        _chunk_queues.emplace_back(std::move(ChunkQueue()));
+    }
+}
+
+Status DataStreamRecvr::PipelineSenderQueue::get_chunk(vectorized::Chunk** chunk, const int32_t driver_sequence) {
+    if (_is_cancelled.load()) {
+        return Status::Cancelled("Cancelled SenderQueueForPipeline::get_chunk");
+    }
+
+    auto& chunk_queue = _is_pipeline_level_shuffle ? _chunk_queues[driver_sequence] : _chunk_queues[0];
+
+    if (chunk_queue.empty()) {
+        return Status::OK();
+    }
+    ChunkItem item;
+    if (!chunk_queue.try_dequeue(item)) {
```

Comment:
There may be race condition between `get_chunk()` and `cancel()`. For example, 
1. `get_chunk()` finds `_is_cancelled` is false and is continue.
2. `cancel()` clears `_chunk_queues`.
3. `get_chunk()` gets `_chunk_queues[driver_sequence]` and exceeds index boundary.

Reply:
> There may be race condition between `get_chunk()` and `cancel()`. For example,
> 
> 1. `get_chunk()` finds `_is_cancelled` is false and is continue.
> 2. `cancel()` clears `_chunk_queues`.
> 3. `get_chunk()` accesses `_chunk_queues[driver_sequence]` and exceeds index boundary.

yes, you're right, I think we can avoid it by clearing _chunk_queues in desctructor instead of cancel